__tests__/ship.test.js

```diff
@@ -0,0 +1,293 @@
+            const emptyConfig = {
+                _itinerary: {
+                    ports: []
+                }
+            };
+            const balticExplorer = new Ship(emptyConfig);
+
+            expect(balticExplorer._isInitialized).toBe(false);
+        });
+    });
+
+    describe('_setSail', () => {
+        let balticExplorerConfig;
+        let balticExplorer;
+        beforeEach(() => {
+            balticExplorerConfig = {
+                _itinerary: {
+                    ports: [
+                        {
+                            name: 'Helsinki',
+                            _ships: [],
+                            addShip: jest.fn(),
+                            removeShip: jest.fn()
+                        },
+                        {
+                            name: 'St Petersburg',
+                            _ships: [],
+                            addShip: jest.fn(),
+                            removeShip: jest.fn()
+                        }
+                    ]
+                }
+            };
+            balticExplorer = new Ship(balticExplorerConfig);
+        });
+
+        afterEach(() => {
+            balticExplorer = null;
+        });
+
+        test('Throws an error if "_itinerary.ports" is empty', () => {
+            const error = 'The ship has no itinerary.';
+
+            balticExplorer._itinerary.ports = [];
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('Throws an error if "_currentPort" is null', () => {
+            const error = 'The ship has already set sail.';
+
+            balticExplorer._currentPort = null;
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('Throws an error if "_itinerary.ports" has one port AND "_previouslyDockedPorts" is not empty', () => {
+            const error = 'The ship has already set sail.';
+
+            balticExplorer._itinerary.ports = [jest.fn()];
+            balticExplorer._previouslyDockedPorts.push(jest.fn());
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('State of "_currentPort" should be set to null', () => {
+            expect(balticExplorer._currentPort).not.toBeNull();
+            balticExplorer._setSail();
+            expect(balticExplorer._currentPort).toBeNull();
+        });
+
+        test('Remove one port from "_itinerary.ports"', () => {
+            expect(balticExplorer._itinerary.ports.length).toBe(2);
+            expect(balticExplorer._itinerary.ports[0].name).toBe('Helsinki');
+            expect(balticExplorer._itinerary.ports[1].name).toBe('St Petersburg');
+            balticExplorer._setSail();
+            expect(balticExplorer._itinerary.ports.length).toBe(1);
+            expect(balticExplorer._itinerary.ports[0].name).toBe('St Petersburg');
```

Comment:
We know the length is going to be 2, and we also know the names of the ports so the first three lines here are redundant